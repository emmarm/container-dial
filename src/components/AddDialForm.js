import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as actions from '../actions';
import getFavicon from '../utils/getFavicon';

class AddDialForm extends Component {
  state = {
    siteName: '',
    siteUrl: ''
  };

  onSiteNameChange = e => {
    const siteName = e.target.value;
    this.setState(() => ({ siteName }));
  };

  onSiteUrlChange = e => {
    const siteUrl = e.target.value;
    this.setState(() => ({ siteUrl }));
  };

  handleSubmit = async e => {
    e.preventDefault();
    const { siteName, siteUrl } = this.state;
    const { container, addDial, handleHideAddDialModal } = this.props;
    const favicon = await getFavicon(siteUrl);
    const dial = {
      siteName,
      siteUrl,
      container: container.name,
      favicon
    };

    addDial(dial);
    handleHideAddDialModal(e);
  };

  render() {
    return (
      <form className="add-dial-modal__form">
        <div className="add-dial-modal__field">
          <label htmlFor="site-name" className="field__label">
            Site Name
          </label>
          <input
            id="site-name"
            className="field__input"
            type="text"
            placeholder="e.g. Facebook"
            onChange={this.onSiteNameChange}
            value={this.state.siteName}
          />
        </div>

        <div className="add-dial-modal__field">
          <label htmlFor="site-url" className="field__label">
            Site URL
          </label>
          <input
            id="site-url"
            className="field__input"
            type="text"
            placeholder="e.g. https://facebook.com"
            onChange={this.onSiteUrlChange}
            value={this.state.siteUrl}
          />
        </div>

        <div className="add-dial-modal__buttons">
          <button
            onClick={this.props.handleHideAddDialModal}
            className="add-dial-modal__button button--secondary"
          >
            Cancel
          </button>

          <button
            className="add-dial-modal__button button--primary"
            onClick={this.handleSubmit}
            style={{ backgroundColor: this.props.theme.primary }}
          >
            Add Dial
          </button>
        </div>
      </form>
    );
  }
}

AddDialForm.propTypes = {
  handleHideAddDialModal: PropTypes.func.isRequired,
  theme: PropTypes.objectOf(PropTypes.string)
};

const mapStateToProps = state => ({
  container: state.container
});

export default connect(
  mapStateToProps,
  actions
)(AddDialForm);
