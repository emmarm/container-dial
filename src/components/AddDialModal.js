import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

const AddDialModal = ({ handleHideAddDialModal, container, isOpen, theme }) => (
  <Modal
    isOpen={isOpen}
    onRequestClose={handleHideAddDialModal}
    closeTimeoutMS={150}
    className="add-dial-modal"
  >
    <h2 className="add-dial-modal__title">
      New{' '}
      <span className="title__span" style={{ color: theme.primary }}>
        {container.name}
      </span>{' '}
      Dial
    </h2>

    <form className="add-dial-modal__form">
      <div className="add-dial-modal__field">
        <label htmlFor="site-name" className="field__label">
          Site Name
        </label>
        <input
          id="site-name"
          className="field__input"
          type="text"
          placeholder="e.g. Facebook"
        />
      </div>

      <div className="add-dial-modal__field">
        <label htmlFor="site-url" className="field__label">
          Site URL
        </label>
        <input
          id="site-url"
          className="field__input"
          type="text"
          placeholder="e.g. https://facebook.com"
        />
      </div>

      <div className="add-dial-modal__buttons">
        <button
          onClick={handleHideAddDialModal}
          className="add-dial-modal__button button--secondary"
        >
          Cancel
        </button>

        <button
          className="add-dial-modal__button button--primary"
          style={{ backgroundColor: theme.primary }}
        >
          Add Dial
        </button>
      </div>
    </form>
  </Modal>
);

AddDialModal.propTypes = {
  handleHideAddDialModal: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  container: PropTypes.objectOf(PropTypes.string)
};

export default AddDialModal;
