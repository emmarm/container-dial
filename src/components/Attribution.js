import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const Text = styled('p')({
  color: 'white',
  fontSize: 14,
  fontStyle: 'italic',
  fontWeight: 100,
  marginRight: 40,
  textAlign: 'right'
});

const Link = styled('a')(
  {
    textDecoration: 'none'
  },
  ({ theme }) => ({
    color: theme.dark
  })
);

const Attribution = ({ image }) => {
  const { user } = !!image && image;
  return (
    <Text>
      Photo by{' '}
      <Link
        href={`${user &&
          user.links.html}?utm_source=container_dial&utm_medium=referral`}
      >
        {user && user.first_name} {user && user.last_name}
      </Link>{' '}
      on{' '}
      <Link href="https://unsplash.com/?utm_source=container_dial&utm_medium=referral">
        Unsplash
      </Link>
    </Text>
  );
};

Attribution.propTypes = {
  image: PropTypes.object
};

export default Attribution;
