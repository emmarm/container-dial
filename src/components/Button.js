import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

import Icon from './Icon';

const Btn = styled('button')(
  {
    alignItems: 'center',
    border: 'none',
    borderRadius: 4,
    boxShadow:
      'rgba(50, 50, 93, 0.1) 0px 7px 14px, rgba(0, 0, 0, 0.08) 0px 3px 6px',
    color: 'white',
    cursor: 'pointer',
    display: 'flex',
    fontSize: 20,
    fontWeight: 100,
    height: 50,
    justifyContent: 'center',
    padding: '0 25px',
    ':disabled': {
      backgroundColor: '#777',
      cursor: 'auto'
    }
  },
  ({ danger, narrow, padded, primary, theme }) => ({
    backgroundColor: primary ? theme.primary : danger ? '#FF3762' : '#bbb',
    padding: narrow && '0 15px',
    paddingLeft: padded && 30,
    ':hover': {
      backgroundColor: primary ? theme.light : danger ? '#FF5F81' : '#c8c8c8',
      ':disabled': {
        backgroundColor: '#777'
      }
    }
  })
);

const Button = ({
  className,
  danger,
  disabled,
  icon,
  narrow,
  onClick,
  primary,
  submit,
  text
}) => (
  <Btn
    className={className}
    danger={danger}
    disabled={disabled}
    onClick={onClick}
    narrow={narrow}
    padded={!!icon && !!text}
    primary={primary}
    type={submit ? 'submit' : 'button'}
  >
    {text}
    {icon && <Icon icon={icon} padded={!!text} />}
  </Btn>
);

Button.propTypes = {
  className: PropTypes.string,
  danger: PropTypes.bool,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  onClick: PropTypes.func,
  primary: PropTypes.bool,
  submit: PropTypes.bool,
  text: PropTypes.string
};

export default Button;
