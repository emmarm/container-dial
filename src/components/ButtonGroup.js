import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const Group = styled('div')({
  display: 'flex',
  justifyContent: 'space-around',
  margin: '25px'
});

const ButtonGroup = ({ children }) => <Group>{children}</Group>;

ButtonGroup.propTypes = {
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.object])
};

export default ButtonGroup;
