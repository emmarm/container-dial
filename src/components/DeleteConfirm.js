import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as actions from '../actions';

export const DeleteConfirm = ({
  dial,
  toggleShowDeleteConfirm,
  handleDelete,
  container,
  theme
}) => (
  <div>
    <p>
      Delete {dial.siteName} from {container.name}?
    </p>

    <button onClick={toggleShowDeleteConfirm}>Cancel</button>

    <button onClick={handleDelete}>Delete</button>
  </div>
);

DeleteConfirm.propTypes = {
  dial: PropTypes.objectOf(PropTypes.string),
  toggleShowDeleteConfirm: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  container: PropTypes.objectOf(PropTypes.string),
  theme: PropTypes.objectOf(PropTypes.string)
};

const mapStateToProps = state => ({
  container: state.container,
  theme: state.theme
});

export default connect(
  mapStateToProps,
  actions
)(DeleteConfirm);
