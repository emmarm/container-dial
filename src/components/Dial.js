import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const DialButton = styled('button')({
  background: 'rgba(255, 255, 255, 0.9)',
  border: 'none',
  boxShadow:
    'rgba(50, 50, 93, 0.1) 0px 7px 14px, rgba(0, 0, 0, 0.08) 0px 3px 6px',
  cursor: 'pointer',
  display: 'grid',
  gridTemplateColumns: '1fr 80px',
  height: '100%',
  padding: 0,
  textDecoration: 'none',
  width: '100%',
  ':hover,:focus': {
    background: 'white'
  },
  '::-moz-focus-inner': {
    border: 0
  }
});

const TitleArea = styled('div')({
  alignItems: 'center',
  color: 'rgba(0,0,0,0.7)',
  display: 'flex',
  fontSize: 18,
  fontWeight: 400,
  height: 80,
  justifyContent: 'center',
  overflow: 'hidden',
  padding: 10,
  textAlign: 'center',
  wordBreak: 'break-word'
});

const Favicon = styled('div')(
  {
    height: 80,
    width: 80
  },
  ({ favicon, theme, useFavicon }) => ({
    background: useFavicon && `center / contain no-repeat url('${favicon}')`,
    color: theme.dark,
    fontSize: 70,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  })
);

class Dial extends Component {
  state = { useFavicon: true };

  async componentDidMount() {
    const { favicon } = this.props.dial;

    const image = new Image();
    image.src = favicon;
    await image.complete;
    image.naturalHeight === 0 && this.setState(() => ({ useFavicon: false }));
  }

  getTrimmedName = siteName => {
    const lastWordEnd = siteName.lastIndexOf(' ', 33);
    return lastWordEnd > -1 && lastWordEnd < 33
      ? siteName.slice(0, lastWordEnd) + '...'
      : siteName.slice(0, 11) + '...';
  };

  getLetter = siteName => siteName.charAt(0).toUpperCase();

  openLink = () => {
    window.location = this.props.dial.siteUrl;
  };

  render() {
    const {
      dial: { favicon, siteName }
    } = this.props;
    const { useFavicon } = this.state;
    return (
      <DialButton onClick={this.openLink} tabIndex={1}>
        <TitleArea>
          {siteName.length > 36 ? this.getTrimmedName(siteName) : siteName}
        </TitleArea>
        <Favicon favicon={favicon} useFavicon={useFavicon}>
          {!useFavicon && this.getLetter(siteName)}
        </Favicon>
      </DialButton>
    );
  }
}

Dial.propTypes = {
  dial: PropTypes.shape({
    container: PropTypes.string,
    favicon: PropTypes.string,
    id: PropTypes.number,
    siteName: PropTypes.string,
    siteUrl: PropTypes.string
  })
};

export default Dial;
