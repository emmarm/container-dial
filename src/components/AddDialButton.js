import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

import Button from './Button';

const AddButton = styled(Button)({
  background: 'rgba(255, 255, 255, 0.3)',
  borderRadius: 0,
  height: '100%',
  ':hover': {
    background: 'rgba(255, 255, 255, 0.5)'
  }
});

const AddDialButton = ({ handleShowDialModal }) => (
  <AddButton onClick={handleShowDialModal} text="Add New" icon="add" />
);

AddDialButton.propTypes = {
  handleShowDialModal: PropTypes.func.isRequired
};

export default AddDialButton;
