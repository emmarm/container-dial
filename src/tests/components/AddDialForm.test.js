import React from 'react';
import { shallow } from 'enzyme';

import { AddDialForm } from '../../components/AddDialForm';

describe('AddDialForm component', () => {
  const props = {
    handleHideAddDialModal: jest.fn(),
    theme: {
      primary: '#333',
      dark: '#000',
      light: '#fff'
    },
    container: {
      name: 'Personal',
      color: 'orange'
    },
    addDial: () => {}
  };
  const wrapper = shallow(<AddDialForm {...props} />);

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('site-name input', () => {
    const nameInput = wrapper.find('#site-name');

    it('updates siteName on change', () => {
      const value = 'Goodreads';
      expect(nameInput.length).toBe(1);
      nameInput.simulate('change', {
        target: { value }
      });
      expect(wrapper.state('siteName')).toBe(value);
    });

    it('sets nameError for empty site name', () => {
      nameInput.simulate('blur', {
        target: { value: '' }
      });
      expect(wrapper.state('nameError')).toBe('Required');
    });

    it('does not set nameError for valid site name', () => {
      nameInput.simulate('blur', {
        target: { value: 'Goodreads' }
      });
      expect(wrapper.state('nameError')).toBe('');
    });
  });

  describe('site-url input', () => {
    const urlInput = wrapper.find('#site-url');

    it('updates siteUrl on change', () => {
      const value = 'https://goodreads.com';
      expect(urlInput.length).toBe(1);
      urlInput.simulate('change', {
        target: { value }
      });
      expect(wrapper.state('siteUrl')).toBe(value);
    });

    it('sets urlError for empty url', () => {
      expect(urlInput.length).toBe(1);
      urlInput.simulate('blur', {
        target: { value: '' }
      });
      expect(wrapper.state('urlError')).toBe('Required');
    });

    it('sets urlError for invalid url', () => {
      urlInput.simulate('blur', {
        target: { value: '@?!' }
      });
      expect(wrapper.state('urlError')).toBe('Invalid URL');
    });

    it('does not set urlError for valid url', () => {
      urlInput.simulate('blur', {
        target: { value: 'https://goodreads.com' }
      });
      expect(wrapper.state('urlError')).toBe('');
    });
  });
});
