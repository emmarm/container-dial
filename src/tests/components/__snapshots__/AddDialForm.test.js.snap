// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDialForm component matches snapshot 1`] = `
<form
  className="add-dial-modal__form"
  onSubmit={[Function]}
>
  <div
    className="add-dial-modal__field"
  >
    <label
      className="field__label"
      htmlFor="site-name"
    >
      Site Name
    </label>
    <input
      className="field__input"
      id="site-name"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="e.g. Facebook"
      type="text"
      value=""
    />
  </div>
  <div
    className="add-dial-modal__field"
  >
    <label
      className="field__label"
      htmlFor="site-url"
    >
      Site URL
    </label>
    <input
      className="field__input"
      id="site-url"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="e.g. https://facebook.com"
      type="text"
      value=""
    />
  </div>
  <div
    className="add-dial-modal__buttons"
  >
    <button
      className="add-dial-modal__button button--secondary"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="add-dial-modal__button button--primary"
      disabled={true}
      style={
        Object {
          "backgroundColor": "#333",
        }
      }
    >
      Add Dial
    </button>
  </div>
</form>
`;
